File: kurrent_skills/templates/golang/projection.go
// KurrentDB Go In-Memory Projection Example
// Demonstrates: Subscribe to events, build state, track checkpoint
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"

	"github.com/google/uuid"
	"github.com/kurrent-io/KurrentDB-Client-Go/kurrentdb"
)

// === MINIMAL PROJECTION FRAMEWORK ===

type EventHandler func(state map[string]interface{}, data map[string]interface{}) map[string]interface{}

type Projection struct {
	Name       string
	State      map[string]map[string]interface{}
	Checkpoint *kurrentdb.Position
	handlers   map[string]EventHandler
}

func NewProjection(name string) *Projection {
	return &Projection{
		Name:     name,
		State:    make(map[string]map[string]interface{}),
		handlers: make(map[string]EventHandler),
	}
}

func (p *Projection) On(eventType string, handler EventHandler) *Projection {
	p.handlers[eventType] = handler
	return p
}

func (p *Projection) Get(streamID string) map[string]interface{} {
	return p.State[streamID]
}

func (p *Projection) Apply(event *kurrentdb.RecordedEvent, position kurrentdb.Position) bool {
	handler, ok := p.handlers[event.EventType]
	if !ok {
		return false
	}

	streamID := event.StreamID
	current := p.State[streamID]
	if current == nil {
		current = make(map[string]interface{})
	}

	var data map[string]interface{}
	json.Unmarshal(event.Data, &data)

	p.State[streamID] = handler(current, data)
	p.Checkpoint = &position
	return true
}

// === ORDER EVENTS (for projection) ===

type ProjectionOrderCreated struct {
	OrderID    string  `json:"orderId"`
	CustomerID string  `json:"customerId"`
	Amount     float64 `json:"amount"`
}

type ProjectionItemAdded struct {
	Item  string  `json:"item"`
	Price float64 `json:"price"`
}

type ProjectionOrderShipped struct {
	ShippedAt string `json:"shippedAt"`
}

// RunProjection runs the in-memory projection example
func RunProjection() {
	ctx := context.Background()

	// === SETUP ===
	connectionString := os.Getenv("KURRENTDB_CONNECTION_STRING")
	if connectionString == "" {
		connectionString = "esdb://localhost:2113?tls=false"
	}

	settings, err := kurrentdb.ParseConnectionString(connectionString)
	if err != nil {
		panic(err)
	}

	client, err := kurrentdb.NewClient(settings)
	if err != nil {
		panic(err)
	}
	defer client.Close()

	fmt.Printf("Connected to KurrentDB at %s\n", connectionString)

	// === DEFINE PROJECTION ===
	orderProjection := NewProjection("OrderSummary").
		On("OrderCreated", func(state map[string]interface{}, data map[string]interface{}) map[string]interface{} {
			return map[string]interface{}{
				"orderId":    data["orderId"],
				"customerId": data["customerId"],
				"amount":     data["amount"],
				"status":     "created",
				"items":      []string{},
			}
		}).
		On("ItemAdded", func(state map[string]interface{}, data map[string]interface{}) map[string]interface{} {
			items := state["items"].([]string)
			amount := state["amount"].(float64)
			state["items"] = append(items, data["item"].(string))
			state["amount"] = amount + data["price"].(float64)
			return state
		}).
		On("OrderShipped", func(state map[string]interface{}, data map[string]interface{}) map[string]interface{} {
			state["status"] = "shipped"
			state["shippedAt"] = data["shippedAt"]
			return state
		}).
		On("OrderCompleted", func(state map[string]interface{}, data map[string]interface{}) map[string]interface{} {
			state["status"] = "completed"
			return state
		})

	// === TEST: Append test events ===
	fmt.Println("\n=== Appending test events ===")

	orderId1 := uuid.New().String()
	orderId2 := uuid.New().String()
	stream1 := fmt.Sprintf("order-%s", orderId1)
	stream2 := fmt.Sprintf("order-%s", orderId2)

	makeEvent := func(eventType string, data interface{}) kurrentdb.EventData {
		jsonData, _ := json.Marshal(data)
		return kurrentdb.EventData{
			EventID:     uuid.New(),
			ContentType: kurrentdb.ContentTypeJson,
			EventType:   eventType,
			Data:        jsonData,
		}
	}

	// Order 1: Created -> ItemAdded -> Shipped -> Completed
	client.AppendToStream(ctx, stream1, kurrentdb.AppendToStreamOptions{},
		makeEvent("OrderCreated", ProjectionOrderCreated{OrderID: orderId1, CustomerID: "cust-1", Amount: 100}))
	client.AppendToStream(ctx, stream1, kurrentdb.AppendToStreamOptions{},
		makeEvent("ItemAdded", ProjectionItemAdded{Item: "Widget", Price: 25}))
	client.AppendToStream(ctx, stream1, kurrentdb.AppendToStreamOptions{},
		makeEvent("OrderShipped", ProjectionOrderShipped{ShippedAt: "2024-01-15T10:00:00Z"}))
	client.AppendToStream(ctx, stream1, kurrentdb.AppendToStreamOptions{},
		makeEvent("OrderCompleted", struct{}{}))

	// Order 2: Created -> ItemAdded (still pending)
	client.AppendToStream(ctx, stream2, kurrentdb.AppendToStreamOptions{},
		makeEvent("OrderCreated", ProjectionOrderCreated{OrderID: orderId2, CustomerID: "cust-2", Amount: 50}))
	client.AppendToStream(ctx, stream2, kurrentdb.AppendToStreamOptions{},
		makeEvent("ItemAdded", ProjectionItemAdded{Item: "Gadget", Price: 30}))

	fmt.Printf("Created order streams: %s, %s\n", stream1, stream2)

	// === RUN PROJECTION ===
	fmt.Println("\n=== Running projection ===")

	subscription, err := client.SubscribeToAll(ctx, kurrentdb.SubscribeToAllOptions{
		From:   kurrentdb.Start{},
		Filter: kurrentdb.ExcludeSystemEventsFilter(),
	})
	if err != nil {
		panic(err)
	}

	processedCount := 0
	targetStreams := map[string]bool{stream1: true, stream2: true}
	targetEventsCount := map[string]int{stream1: 0, stream2: 0}
	expectedCounts := map[string]int{stream1: 4, stream2: 2} // Order 1: 4 events, Order 2: 2 events

	for {
		event := subscription.Recv()

		if event.SubscriptionDropped != nil {
			break
		}

		if event.EventAppeared != nil {
			evt := event.EventAppeared.OriginalEvent()
			position := event.EventAppeared.OriginalEvent().Position

			if orderProjection.Apply(evt, position) {
				processedCount++
				fmt.Printf("  Processed: %s on %s\n", evt.EventType, evt.StreamID)

				if targetStreams[evt.StreamID] {
					targetEventsCount[evt.StreamID]++
				}
			}

			// Stop when we've processed all our test events for both streams
			if targetEventsCount[stream1] >= expectedCounts[stream1] &&
				targetEventsCount[stream2] >= expectedCounts[stream2] {
				break
			}

			// Safety limit
			if processedCount > 200 {
				break
			}
		}
	}
	subscription.Close()

	// === VERIFY RESULTS ===
	fmt.Println("\n=== Projection Results ===")

	order1State := orderProjection.Get(stream1)
	order2State := orderProjection.Get(stream2)

	order1JSON, _ := json.MarshalIndent(order1State, "", "  ")
	order2JSON, _ := json.MarshalIndent(order2State, "", "  ")

	fmt.Printf("\nOrder 1 (%s):\n%s\n", stream1, string(order1JSON))
	fmt.Printf("\nOrder 2 (%s):\n%s\n", stream2, string(order2JSON))

	// === ASSERTIONS ===
	fmt.Println("\n=== Running assertions ===")

	passed := true

	// Order 1 assertions
	if order1State["status"] != "completed" {
		fmt.Printf("FAIL: Order 1 status should be 'completed', got '%v'\n", order1State["status"])
		passed = false
	}
	if order1State["amount"].(float64) != 125 {
		fmt.Printf("FAIL: Order 1 amount should be 125, got %v\n", order1State["amount"])
		passed = false
	}
	if len(order1State["items"].([]string)) != 1 {
		fmt.Printf("FAIL: Order 1 should have 1 item, got %d\n", len(order1State["items"].([]string)))
		passed = false
	}

	// Order 2 assertions
	if order2State["status"] != "created" {
		fmt.Printf("FAIL: Order 2 status should be 'created', got '%v'\n", order2State["status"])
		passed = false
	}
	if order2State["amount"].(float64) != 80 {
		fmt.Printf("FAIL: Order 2 amount should be 80, got %v\n", order2State["amount"])
		passed = false
	}

	// Checkpoint assertion
	if orderProjection.Checkpoint == nil {
		fmt.Println("FAIL: Checkpoint should be set")
		passed = false
	}

	if passed {
		fmt.Println("\nAll projection tests passed!")
	} else {
		fmt.Println("\nSome tests failed!")
		os.Exit(1)
	}
}

// Uncomment to run as standalone:
// func main() {
// 	runProjection()
// }

File: kurrent_skills/templates/python/projection.py
#!/usr/bin/env python3
"""
KurrentDB Python In-Memory Projection Example
Demonstrates: Subscribe to events, build state, track checkpoint
"""

import json
import os
import sys
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Optional
from kurrentdbclient import KurrentDBClient, NewEvent, StreamState


def log(msg):
    print(msg)
    sys.stdout.flush()


# === MINIMAL PROJECTION FRAMEWORK ===
@dataclass
class Projection:
    name: str
    state: Dict[str, Any] = field(default_factory=dict)
    checkpoint: Optional[int] = None
    _handlers: Dict[str, Callable] = field(default_factory=dict)

    def on(self, event_type: str, handler: Callable) -> "Projection":
        self._handlers[event_type] = handler
        return self

    def get(self, stream_id: str) -> Optional[Any]:
        return self.state.get(stream_id)

    def get_all(self) -> Dict[str, Any]:
        return self.state.copy()

    def apply(self, event) -> bool:
        handler = self._handlers.get(event.type)
        if handler:
            stream_id = event.stream_name
            current = self.state.get(stream_id, {})
            data = json.loads(event.data.decode()) if event.data else {}
            self.state[stream_id] = handler(current, data, event)
            self.checkpoint = event.commit_position
            return True
        return False


# === SETUP ===
connection_string = os.environ.get(
    "KURRENTDB_CONNECTION_STRING",
    "kurrentdb://localhost:2113?tls=false"
)
client = KurrentDBClient(uri=connection_string)
log(f"Connected to KurrentDB at {connection_string}")


# === DEFINE PROJECTION ===
def handle_order_created(state, data, event):
    return {
        "order_id": data.get("orderId"),
        "customer_id": data.get("customerId"),
        "amount": data.get("amount"),
        "status": "created",
        "items": []
    }


def handle_item_added(state, data, event):
    return {
        **state,
        "items": state.get("items", []) + [data.get("item")],
        "amount": state.get("amount", 0) + data.get("price", 0)
    }


def handle_order_shipped(state, data, event):
    return {
        **state,
        "status": "shipped",
        "shipped_at": data.get("shippedAt")
    }


def handle_order_completed(state, data, event):
    return {
        **state,
        "status": "completed"
    }


order_projection = Projection("OrderSummary")
order_projection.on("OrderCreated", handle_order_created)
order_projection.on("ItemAdded", handle_item_added)
order_projection.on("OrderShipped", handle_order_shipped)
order_projection.on("OrderCompleted", handle_order_completed)


# === TEST: Append test events ===
log("\n=== Appending test events ===")

order_id_1 = str(uuid.uuid4())
order_id_2 = str(uuid.uuid4())
stream_1 = f"order-{order_id_1}"
stream_2 = f"order-{order_id_2}"


def make_event(event_type: str, data: dict) -> NewEvent:
    return NewEvent(
        type=event_type,
        data=json.dumps(data).encode("utf-8"),
        content_type="application/json"
    )


# Order 1: Created -> ItemAdded -> Shipped -> Completed
client.append_to_stream(
    stream_name=stream_1,
    current_version=StreamState.ANY,
    events=[make_event("OrderCreated", {"orderId": order_id_1, "customerId": "cust-1", "amount": 100})]
)
client.append_to_stream(
    stream_name=stream_1,
    current_version=StreamState.ANY,
    events=[make_event("ItemAdded", {"item": "Widget", "price": 25})]
)
client.append_to_stream(
    stream_name=stream_1,
    current_version=StreamState.ANY,
    events=[make_event("OrderShipped", {"shippedAt": "2024-01-15T10:00:00Z"})]
)
client.append_to_stream(
    stream_name=stream_1,
    current_version=StreamState.ANY,
    events=[make_event("OrderCompleted", {})]
)

# Order 2: Created -> ItemAdded (still pending)
client.append_to_stream(
    stream_name=stream_2,
    current_version=StreamState.ANY,
    events=[make_event("OrderCreated", {"orderId": order_id_2, "customerId": "cust-2", "amount": 50})]
)
client.append_to_stream(
    stream_name=stream_2,
    current_version=StreamState.ANY,
    events=[make_event("ItemAdded", {"item": "Gadget", "price": 30})]
)

log(f"Created order streams: {stream_1}, {stream_2}")


# === RUN PROJECTION ===
log("\n=== Running projection ===")

subscription = client.subscribe_to_all(commit_position=0)

target_streams = {stream_1, stream_2}
target_events_count = {stream_1: 0, stream_2: 0}
expected_counts = {stream_1: 4, stream_2: 2}  # Order 1: 4 events, Order 2: 2 events

for event in subscription:
    # Skip system streams
    if event.stream_name.startswith("$"):
        continue

    if order_projection.apply(event):
        log(f"  Processed: {event.type} on {event.stream_name}")

        if event.stream_name in target_streams:
            target_events_count[event.stream_name] += 1

    # Stop when we've processed all our test events
    if (target_events_count[stream_1] >= expected_counts[stream_1] and
        target_events_count[stream_2] >= expected_counts[stream_2]):
        break

    # Safety limit
    total_processed = sum(target_events_count.values())
    if total_processed > 200:
        break


# === VERIFY RESULTS ===
log("\n=== Projection Results ===")

order1_state = order_projection.get(stream_1)
order2_state = order_projection.get(stream_2)

log(f"\nOrder 1 ({stream_1}):")
log(json.dumps(order1_state, indent=2))

log(f"\nOrder 2 ({stream_2}):")
log(json.dumps(order2_state, indent=2))


# === ASSERTIONS ===
log("\n=== Running assertions ===")

passed = True

# Order 1 assertions
if order1_state.get("status") != "completed":
    log(f"FAIL: Order 1 status should be 'completed', got '{order1_state.get('status')}'")
    passed = False

if order1_state.get("amount") != 125:
    log(f"FAIL: Order 1 amount should be 125, got {order1_state.get('amount')}")
    passed = False

if len(order1_state.get("items", [])) != 1:
    log(f"FAIL: Order 1 should have 1 item, got {len(order1_state.get('items', []))}")
    passed = False

# Order 2 assertions
if order2_state.get("status") != "created":
    log(f"FAIL: Order 2 status should be 'created', got '{order2_state.get('status')}'")
    passed = False

if order2_state.get("amount") != 80:
    log(f"FAIL: Order 2 amount should be 80, got {order2_state.get('amount')}")
    passed = False

# Checkpoint assertion
if order_projection.checkpoint is None:
    log("FAIL: Checkpoint should be set")
    passed = False

if passed:
    log("\nAll projection tests passed!")
else:
    log("\nSome tests failed!")
    sys.exit(1)

client.close()
